### PR TITLE
kanidm: Fix website and pam/nss in package

### PR DESCRIPTION
**Pull request by Flakebi**
Also fix some bugs in the module and improve the test

### CODE REVIEW

File: nixos/modules/services/security/kanidm.nix [hunk 1]
```diff
@@ -199,10 +200,32 @@ in
...
+            echo "Using temporary config with old domain name '$olddomain'"
+            # Create temporary config with old domain name
+            config="$(${pkgs.coreutils}/bin/mktemp)"
+            ${pkgs.gnused}/bin/sed ${dummyServerConfigFile} -e "s#@origin@#https://$olddomain#" > "$config"
+            ${pkgs.kanidm}/bin/kanidmd domain_name_change \
+              -c "$config" \
+              -n ${cfg.ensureDomainName}
+            ${pkgs.coreutils}/bin/rm "$config"
+          fi
         '';
```
Comment (mweinelt) [inline nixos/modules/services/security/kanidm.nix:238]: Using `systemd.services.kanidm.path = with pkgs; [ coreutils  gnugrep gnused ];` would make this script look less mutilated! :laughing:

File: nixos/tests/kanidm.nix [hunk 2]
```diff
@@ -16,24 +12,41 @@ import ./make-test-python.nix ({ pkgs, ... }:
...
+              proxy_cache_bypass  $http_upgrade;
+
+              proxy_set_header Upgrade           $http_upgrade;
+              proxy_set_header Connection        "upgrade";
+              proxy_set_header Host              $host;
+              proxy_set_header X-Real-IP         $remote_addr;
+              proxy_set_header X-Forwarded-For   $proxy_add_x_forwarded_for;
+              proxy_set_header X-Forwarded-Proto $scheme;
+              proxy_set_header X-Forwarded-Host  $host;
+              proxy_set_header X-Forwarded-Port  $server_port;
```
Comment (mweinelt) [inline nixos/tests/kanidm.nix]: These lines boil down to `recommendedProxySettings` and `proxyWebsockets`.

File: nixos/tests/kanidm.nix [hunk 3]
```diff
@@ -42,17 +55,17 @@ import ./make-test-python.nix ({ pkgs, ... }:
 
       environment.etc."kanidm/config".text = ''
         uri = "https://${serverDomain}"
-        verify_ca = true
-        verify_hostnames = true
       '';
```
Comment (mweinelt) [inline nixos/tests/kanidm.nix]: What does this configure? Shouldn't this part be configured through the kanidm client module bits?

Comment (Flakebi) [reply nixos/tests/kanidm.nix]: There’s not really a client module at this moment, so the config file only gets created if a server or pam support is enabled :/

Oh, and these verify settings are `true` by default, so I removed them.

Comment (mweinelt) [reply nixos/tests/kanidm.nix]: Yeah, so I wonder why we need this in the test instead of letting the module create it.

Comment (Flakebi) [reply nixos/tests/kanidm.nix]: I split the options into `enableClient` and `enableServer`, so the test can use `clientSettings` now.

**Comment by erictapen** [comment]
And again I didn't look properly before working on the branch ;)
I fixed the `ldapbindaddress` issue by filtering null values in 964ec52d954a7d3f666769ae8e32b29223dad550. This way we are able to document the option. Could you rebase ontop of that?

File: nixos/modules/services/security/kanidm.nix [hunk 4]
```diff
@@ -189,20 +198,43 @@ in
...
+
+            echo "Using temporary config with old domain name '$olddomain'"
+            # Create temporary config with old domain name
+            config="$(mktemp)"
+            sed ${dummyServerConfigFile} -e "s#@origin@#https://$olddomain#" > "$config"
+            ${pkgs.kanidm}/bin/kanidmd domain_name_change \
+              -c "$config" \
+              -n ${cfg.ensureDomainName}
+            rm "$config"
+          fi
```
Comment (mweinelt) [inline nixos/modules/services/security/kanidm.nix:237]: Please present that hack upstream so it gets proper recognition. :facepalm: 

Anyway, it works!

**Comment by mweinelt** [comment]
image


:partying_face:

File: nixos/modules/services/security/kanidm.nix (hunk 4 above)
Comment (erictapen) [reply nixos/modules/services/security/kanidm.nix:237]: I'm really hesitant about that amount of shell magic in init files. Would really prefer that logic to be handled by upstream. But as it is not user facing, I'd be okay with it if it works.

Comment (Flakebi) [reply nixos/modules/services/security/kanidm.nix:237]: I can see what you mean :D I mentioned it in https://github.com/kanidm/kanidm/issues/623

**Pull request merged**